File: marriage_decision.m
function [M,prev_state_w,WE,HE,max_weighted_utility_index]=marriage_decision(U_H, U_W, BP, outside_option_h_v,outside_option_w_v,outside_option_w, WE, HE)  
    BP_FLAG_PLUS = 0;
    BP_FLAG_MINUS = 0;
    if (BP<0)
        error()
    end
       looking_for_solution = 1;
       max_iterations = 10;
       M = 0;
       prev_state_w = 0;
        while (looking_for_solution == 1)
            if (max_iterations == 0)
                BP
                BP_FLAG_PLUS
                BP_FLAG_MINUS
            end
            max_iterations = max_iterations - 1;
            weighted_utility=U_H(1:22).*BP+U_W(1:22).*(1-BP);  %weighted utility
            [max_weighted_utility,max_weighted_utility_index]=max(weighted_utility);      % max of weighted utility
           % max_weighted_utility_index
            if (U_H(max_weighted_utility_index) >= outside_option_h_v && U_W(max_weighted_utility_index) >= outside_option_w_v) 
                %the max in married for both is better than outside
                looking_for_solution = 0;
                M = 1 ;
                if (max_weighted_utility_index < 12)   %1-12 WOMEN UNEMPLOYED
                    prev_state_w = 0;
                    WE = WE;
                    HE = HE+1;
                elseif (max_weighted_utility_index > 11)   %12-22: wife employed  + husband employed
                    prev_state_w = 1;
                    WE = WE+1;
                    HE = HE+1;
                else
                    error()
                end            
            elseif (U_H(max_weighted_utility_index)< outside_option_h_v && U_W(max_weighted_utility_index)<outside_option_w_v) 
                %the outside option is better for both - no marriage
                looking_for_solution = 0;
                M = 0 ;
                if (outside_option_w_v == 1)   %3-unmarried+wife unemployed
                    prev_state_w = 0;
                    WE = WE;
                else     %3-unmarried+wife employed
                    prev_state_w = 1;
                    WE = WE+1;
                end    
            elseif (U_H(max_weighted_utility_index) >= outside_option_h_v && U_W(max_weighted_utility_index) < outside_option_w_v) 
                %the outside option is better for women
                % increase the wife bargaining power
                BP=BP-0.1;
                BP_FLAG_MINUS = 1;
                if (BP<0)|| (BP_FLAG_PLUS == 1)%the first condition: no solution the second condition: infinite loop of BP
                    looking_for_solution = 0;
                    M = 0 ;
                    if (outside_option_w == 1)   %3-unmarried+wife unemployed
                        prev_state_w = 0;
                        WE = WE;
                    else     %3-unmarried+wife employed
                        prev_state_w = 1;
                        WE = WE+1;
                    end    
                end    
            elseif (U_H(max_weighted_utility_index) < outside_option_h_v && U_W(max_weighted_utility_index) >= outside_option_w_v) 
                %the outside option is better for husband
                % increase the husband bargaining power
                BP=BP+0.1;
                BP_FLAG_PLUS = 1;
                if (BP>1)||(BP_FLAG_MINUS == 1)%the first condition: no solution the second condition: infinite loop of BP
                    looking_for_solution = 0;
                    M = 0 ;
                    if (outside_option_w == 1)   %3-unmarried+wife unemployed
                        prev_state_w = 0;
                        WE = WE;
                    else     %3-unmarried+wife employed
                        prev_state_w = 1;
                        WE = WE+1;
                    end    
                end    
            else
                error(0)
            end
        end%while - loop over BP, increase/decrease BP 

File: c++/marriage_decision.cpp
#include "calculate_utility.h"
#include "marriage_decision.h"
#include <stdexcept>
#include <string>
#include <cassert>

MarriageDecision marriage_decision(const Utility& utility, double BP, unsigned WE, unsigned HE) {

    MarriageDecision result;
    result.outside_option_w_v = std::max(utility.U_W_S[0], utility.U_W_S[1]);
    result.outside_option_h_v = utility.U_H_S;
    result.outside_option_w = (utility.U_W_S[0] > utility.U_W_S[1]) ? UNEMP : EMP;
    result.M = 0;
    result.prev_state_w = 0;
    
    if (BP < 0) {
        // no marriage is possible to begin with
        return result;
    }

    bool BP_FLAG_PLUS{false};
    bool BP_FLAG_MINUS{false};
    unsigned max_iterations = 10;

    while (true) {
        if (max_iterations == 0) {
            throw std::runtime_error("max iteration reached. BP=" + std::to_string(BP));
        }
        --max_iterations;

        UtilityArray weighted_utility;
        for (auto i = 0U; i < UTILITY_SIZE; ++i) {
            weighted_utility[i] = utility.U_H[i]*BP + utility.U_W[i]*(1.0-BP);
        }

        const auto max_weighted_utility_index = std::max_element(weighted_utility.begin(), weighted_utility.end()) - weighted_utility.begin();
        assert(max_weighted_utility_index < UTILITY_SIZE);
        [[maybe_unused]] const auto max_weighted_utility = weighted_utility[max_weighted_utility_index];
        const auto max_U_W = utility.U_W[max_weighted_utility_index];
        const auto max_U_H = utility.U_H[max_weighted_utility_index];
        result.max_weighted_utility_index = max_weighted_utility_index;

        if (max_U_H >= result.outside_option_h_v && max_U_W >= result.outside_option_w_v) {
            // the max in married for both is better than outside
            result.M = 1 ;
            if (max_weighted_utility_index < 11) {  
                // 0-10 WOMEN UNEMPLOYED
                result.prev_state_w = 0;
                result.WE = WE;
                result.HE = HE+1;
                return result;
            }
            // 11-21: wife employed  + husband employed
            result.prev_state_w = 1;
            result.WE = WE+1;
            result.HE = HE+1;
            return result;
        } else if (max_U_H < result.outside_option_h_v && max_U_H < result.outside_option_w_v) { 
            // the outside option is better for both - no marriage
            break;
        } else if (max_U_H  >= result.outside_option_h_v && max_U_W < result.outside_option_w_v) { 
            // the outside option is better for wife
            // increase the wife bargaining power
            BP -= 0.1;
            BP_FLAG_MINUS = true;
            if (BP < 0.0 || BP_FLAG_PLUS) { 
                // no solution - boundry reached or infinite loop of BP
                break;
            }
        } else if (max_U_H < result.outside_option_h_v && max_U_W >= result.outside_option_w_v) {
            // the outside option is better for husband
            // increase the husband bargaining power
            BP += 0.1;
            BP_FLAG_PLUS = true;
            if (BP > 1 || BP_FLAG_MINUS) {
                // no solution - boundry reached or infinite loop of BP
                break;
            }
        }
    }

    // no marriage
    result.M = 0;
    // TODO: check equality to 1 of double
    if (result.outside_option_w_v == 1) {  
        // 3-unmarried+wife unemployed
        result.prev_state_w = 0;
        result.WE = WE;
        return result;
    }
    // 3-unmarried+wife employed
    result.prev_state_w = 1;
    result.WE = WE+1;
    return result;   
}

